File: backend/app/schemas/employer.py
from datetime import datetime
from typing import Annotated, Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, HttpUrl, field_validator


class EmployerCompanyCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    company_name: Annotated[str, Field(min_length=1, max_length=160)]
    website: HttpUrl | None = None
    description: Annotated[str | None, Field(default=None, max_length=5000)] = None


class EmployerCompanyUpdateRequest(BaseModel):
    """Partial update of existing company fields. Omitted fields stay unchanged."""

    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    company_name: Annotated[str | None, Field(default=None, min_length=1, max_length=160)] = None
    website: HttpUrl | None = None
    description: Annotated[str | None, Field(default=None, max_length=5000)] = None

    @field_validator("company_name", mode="before")
    @classmethod
    def reject_null_company_name(cls, value: object) -> object:
        if value is None:
            raise ValueError("company_name cannot be null")
        return value


class EmployerCompanyResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    company_name: str
    website: str | None
    description: str | None
    created_at: datetime


VacancyStatus = Literal["draft", "open", "closed"]


class VacancyCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid", str_strip_whitespace=True)

    title: Annotated[str, Field(min_length=1, max_length=200)]
    description: Annotated[str | None, Field(default=None, max_length=5000)] = None
    status: VacancyStatus = "open"


class VacancyResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    title: str
    description: str | None
    status: VacancyStatus
    created_at: datetime


VacancyRequirementType = Literal["required", "preferred"]


class VacancyRequirementCreateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    skill_id: UUID
    requirement_type: VacancyRequirementType


class VacancyRequirementResponse(BaseModel):
    id: UUID
    skill_id: UUID
    skill_name: str
    skill_category: str
    requirement_type: VacancyRequirementType


class SkillOptionResponse(BaseModel):
    id: UUID
    name: str
    category: str


class SignalSummaryResponse(BaseModel):
    """Employer-safe public category describing why evidence supports a skill."""

    category: str


class MatchedSkillEvidenceResponse(BaseModel):
    """Employer-safe evidence unit supporting one matched vacancy skill."""

    id: UUID
    source_type: str
    title: str | None
    verification_status: str | None
    ownership_status: str | None
    evidence_confidence: float
    signal_summaries: list[SignalSummaryResponse] = Field(default_factory=list)


class MatchedSkillDetailsResponse(BaseModel):
    skill_id: UUID
    skill_name: str
    evidence: list[MatchedSkillEvidenceResponse]


class EvidenceSuggestionResponse(BaseModel):
    """One public evidence category that could confirm a missing skill."""

    category: str


class MissingSkillDetailsResponse(BaseModel):
    """Deterministic evidence channels for one skill absent from the passport."""

    skill_id: UUID
    skill_name: str
    evidence_suggestions: list[EvidenceSuggestionResponse]


class MatchSkillGroupResponse(BaseModel):
    matched: list[str]
    missing: list[str]
    matched_details: list[MatchedSkillDetailsResponse] = Field(default_factory=list)
    missing_details: list[MissingSkillDetailsResponse] = Field(default_factory=list)


class VacancyMatchResponse(BaseModel):
    candidate_id: UUID
    candidate_name: str
    score: int
    required: MatchSkillGroupResponse
    preferred: MatchSkillGroupResponse


class VacancyMatchesResponse(BaseModel):
    matches: list[VacancyMatchResponse]


EmployerCandidateStage = Literal[
    "shortlisted",
    "screening",
    "interview",
    "offer",
    "hired",
    "rejected",
]


class EmployerShortlistStageUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    stage: EmployerCandidateStage


class EmployerShortlistNoteUpdateRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    note: str | None

    @field_validator("note", mode="before")
    @classmethod
    def normalize_note(cls, value: object) -> str | None:
        if value is None:
            return None
        if not isinstance(value, str):
            raise ValueError("Input should be a valid string")
        stripped = value.strip()
        if stripped == "":
            return None
        if len(stripped) > 5000:
            raise ValueError("String should have at most 5000 characters")
        return stripped


class EmployerShortlistEntryResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    vacancy_id: UUID
    candidate_id: UUID
    stage: EmployerCandidateStage
    note: str | None
    created_at: datetime
    updated_at: datetime


class EmployerShortlistListResponse(BaseModel):
    entries: list[EmployerShortlistEntryResponse]


class MatchDetailsCandidateResponse(BaseModel):
    id: UUID
    name: str
    headline: str | None
    avatar: str | None


class MatchDetailsMatchResponse(BaseModel):
    score: int
    required: MatchSkillGroupResponse
    preferred: MatchSkillGroupResponse


class MatchDetailsPassportSkillResponse(BaseModel):
    """Employer-safe, read-only projection of an existing passport skill."""

    name: str
    evidence_confidence: float
    evidence_count: int
    source_types: list[str]


class MatchDetailsPassportResponse(BaseModel):
    # Retained for existing match-review clients.
    top_skills: list[str]
    # Values come directly from the candidate's existing Skill Passport.
    skills: list[MatchDetailsPassportSkillResponse] = []


class MatchDetailsEvidenceResponse(BaseModel):
    source_type: str
    title: str | None
    verification_status: str | None
    ownership_status: str | None
    skills: list[str]


class MatchDetailsRoadmapItemResponse(BaseModel):
    id: str
    title: str
    reason: str
    priority: Literal["high", "medium", "low"]
    missing_skills: list[str]
    related_skills: list[str]


class MatchDetailsResponse(BaseModel):
    candidate: MatchDetailsCandidateResponse
    match: MatchDetailsMatchResponse
    passport: MatchDetailsPassportResponse
    evidence: list[MatchDetailsEvidenceResponse]
    roadmap: list[MatchDetailsRoadmapItemResponse]
    has_applied: bool = False


ApplicationStatus = Literal["applied", "withdrawn"]


class EmployerApplicantResponse(BaseModel):
    application_id: UUID
    candidate_id: UUID
    candidate_name: str
    status: ApplicationStatus
    applied_at: datetime
    score: int
    required: MatchSkillGroupResponse
    preferred: MatchSkillGroupResponse


class EmployerApplicantsResponse(BaseModel):
    applicants: list[EmployerApplicantResponse]


class ApplicantContactResponse(BaseModel):
    email: str
    phone: str | None
    telegram: str | None
    linkedin_url: str | None
    portfolio_url: str | None
    location: str | None


class AiMatchExplanationResponse(BaseModel):
    summary: str
    strengths: list[str]
    gaps: list[str]
    next_steps: list[str]

File: frontend/lib/api/types/employer.ts
export type EmployerCompany = {
  id: string;
  company_name: string;
  website: string | null;
  description: string | null;
  created_at: string;
};

export type EmployerCompanyCreateRequest = {
  company_name: string;
  website?: string | null;
  description?: string | null;
};

export type EmployerCompanyUpdateRequest = {
  company_name?: string;
  website?: string | null;
  description?: string | null;
};

export type VacancyStatus = "draft" | "open" | "closed";

export type Vacancy = {
  id: string;
  title: string;
  description: string | null;
  status: VacancyStatus;
  created_at: string;
};

export type VacancyCreateRequest = {
  title: string;
  description?: string | null;
  status?: VacancyStatus;
};

export type VacancyRequirementType = "required" | "preferred";

export type VacancyRequirement = {
  id: string;
  skill_id: string;
  skill_name: string;
  skill_category: string;
  requirement_type: VacancyRequirementType;
};

export type VacancyRequirementCreateRequest = {
  skill_id: string;
  requirement_type: VacancyRequirementType;
};

export type SkillOption = {
  id: string;
  name: string;
  category: string;
};

export type SignalSummary = {
  category: string;
};

export type MatchedSkillEvidence = {
  id: string;
  source_type: string;
  title: string | null;
  verification_status: string | null;
  ownership_status: string | null;
  evidence_confidence: number;
  /** Absent on older responses; treat as empty. */
  signal_summaries?: SignalSummary[];
};

export type MatchedSkillDetails = {
  skill_id: string;
  skill_name: string;
  evidence: MatchedSkillEvidence[];
};

export type EvidenceSuggestion = {
  category: string;
};

export type MissingSkillDetails = {
  skill_id: string;
  skill_name: string;
  evidence_suggestions: EvidenceSuggestion[];
};

export type MatchSkillGroup = {
  matched: string[];
  missing: string[];
  /** Absent on older responses / list endpoint; treat as empty. */
  matched_details?: MatchedSkillDetails[];
  /** Absent on older responses / list endpoint; treat as empty. */
  missing_details?: MissingSkillDetails[];
};

export type VacancyMatch = {
  candidate_id: string;
  candidate_name: string;
  score: number;
  required: MatchSkillGroup;
  preferred: MatchSkillGroup;
};

export type VacancyMatchesResponse = {
  matches: VacancyMatch[];
};

export type MatchDetailsCandidate = {
  id: string;
  name: string;
  headline: string | null;
  avatar: string | null;
};

export type MatchDetailsMatch = {
  score: number;
  required: MatchSkillGroup;
  preferred: MatchSkillGroup;
};

export type MatchDetailsPassport = {
  top_skills: string[];
  /** Optional while independently deployed backends may still return the legacy summary. */
  skills?: MatchDetailsPassportSkill[];
};

export type MatchDetailsPassportSkill = {
  name: string;
  evidence_confidence: number;
  evidence_count: number;
  source_types: string[];
};

export type MatchDetailsEvidence = {
  source_type: string;
  title: string | null;
  /** Present on current backends; optional for older fixtures. */
  verification_status?: string | null;
  /** Present on current backends; optional for older fixtures. */
  ownership_status?: string | null;
  skills: string[];
};

export type MatchDetailsRoadmapItem = {
  id: string;
  title: string;
  reason: string;
  priority: "high" | "medium" | "low";
  missing_skills: string[];
  related_skills: string[];
};

export type MatchDetailsResponse = {
  candidate: MatchDetailsCandidate;
  match: MatchDetailsMatch;
  passport: MatchDetailsPassport;
  evidence: MatchDetailsEvidence[];
  roadmap: MatchDetailsRoadmapItem[];
  has_applied?: boolean;
};

export type ApplicationStatus = "applied" | "withdrawn";

export type EmployerApplicant = {
  application_id: string;
  candidate_id: string;
  candidate_name: string;
  status: ApplicationStatus;
  applied_at: string;
  score: number;
  required: MatchSkillGroup;
  preferred: MatchSkillGroup;
};

export type EmployerApplicantsResponse = {
  applicants: EmployerApplicant[];
};

export type ApplicantContact = {
  email: string;
  phone: string | null;
  telegram: string | null;
  linkedin_url: string | null;
  portfolio_url: string | null;
  location: string | null;
};

export type AiMatchExplanation = {
  summary: string;
  strengths: string[];
  gaps: string[];
  next_steps: string[];
};

export type EmployerCandidateStage =
  | "shortlisted"
  | "screening"
  | "interview"
  | "offer"
  | "hired"
  | "rejected";

export const EMPLOYER_CANDIDATE_STAGES = [
  "shortlisted",
  "screening",
  "interview",
  "offer",
  "hired",
  "rejected"
] as const satisfies readonly EmployerCandidateStage[];

export const EMPLOYER_CANDIDATE_STAGE_LABELS: Record<EmployerCandidateStage, string> = {
  shortlisted: "Shortlisted",
  screening: "Screening",
  interview: "Interview",
  offer: "Offer",
  hired: "Hired",
  rejected: "Rejected"
};

export type EmployerShortlistEntry = {
  id: string;
  vacancy_id: string;
  candidate_id: string;
  stage: EmployerCandidateStage;
  note: string | null;
  created_at: string;
  updated_at: string;
};

export type EmployerShortlistNoteUpdateRequest = {
  note: string | null;
};

export type EmployerShortlistResponse = {
  entries: EmployerShortlistEntry[];
};
